src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensRangeHandler.cs

```diff
@@ -76,44 +143,49 @@ public SemanticTokensRangeHandler(IGlobalOptionService globalOptions)
 
-            return new RoslynSemanticTokens { Data = tokensData, IsFinalized = isFinalized };
-        }
-
-        private async Task<bool> IsDataFinalizedAsync(Project project, CancellationToken cancellationToken)
-        {
-            // We use a combination of IsFullyLoaded + the completed project compilation as the metric
-            // for isFinalized. It may not be completely accurate but this is only a a temporary fix until
-            // workspace/semanticTokens/refresh is implemented.
-            lock (_lock)
+            // The above call to get semantic tokens may be inaccurate (because we use frozen partial semantics).  Kick
+            // off a request to ensure that the OOP side gets a fully up to compilation for this project.  Once it does
+            // we can optionally choose to notify our caller to do a refresh if we computed a compilation for a new
+            // solution snapshot.
+            if (_supportsRefresh.Value)
             {
-                if (_projectIdToCompilation.TryGetValue(project.Id, out var compilationTask) && compilationTask.IsCompleted)
+                lock (_gate)
                 {
-                    // We don't want to hang on to the compilation since this can be very expensive,
-                    // but we do want to mark the compilation as being successfully retrieved.
-                    if (compilationTask.Result is not null)
+                    if (!_projectIdToEventSource.TryGetValue(project.Id, out var eventSource))
                     {
-                        _projectIdToCompilation[project.Id] = Task.FromResult<Compilation?>(null);
+                        eventSource = new CompilationAvailableEventSource(_asyncListener);
+                        _projectIdToEventSource.Add(project.Id, eventSource);
                     }
 
-                    return true;
+                    // Kick off work to have this project be sync'ed over to OOP so it's ready in the future for an upcoming 
+                    eventSource.EnsureCompilationAvailability(project, OnCompilationAvailableAsync);
                 }
             }
 
-            var workspaceStatusService = project.Solution.Workspace.Services.GetRequiredService<IWorkspaceStatusService>();
-            var isFullyLoaded = await workspaceStatusService.IsFullyLoadedAsync(cancellationToken).ConfigureAwait(false);
+            return new LSP.SemanticTokens { Data = tokensData };
+        }
 
-            lock (_lock)
+        private async ValueTask OnCompilationAvailableAsync(Project project, CancellationToken cancellationToken)
+        {
+            var solutionChecksum = await project.Solution.State.GetChecksumAsync(cancellationToken).ConfigureAwait(false);
+            lock (_gate)
             {
-                if (isFullyLoaded && !_projectIdToCompilation.ContainsKey(project.Id))
+                if (_projectIdLastComputedSolutionChecksum.TryGetValue(project.Id, out var lastComputedChecksum) &&
```

Comment:
I think this needs to be moved up - this looks like we compute the compilation then afterwards we compare to the last result.  But that means when the checksum is the same we compute the compilation anyway, we just don't store it or send a notification